src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php

```diff
@@ -386,7 +386,8 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
             $definition->addMethodCall('setSecurityInformation', ['%sonata.admin.configuration.security.information%']);
         }
 
-        $definition->addMethodCall('initialize');
+        $definition->addMethodCall('setFormTheme', [$overwriteAdminConfiguration['templates']['form'] ?? []]);
```

Comment:
should we check if the user has already call `setFormTheme` in the service definition and merge the arguments?

Reply:
I would say that if you want to set extra form theme, you should do it with a lower priority instead.

For example https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1427

Reply:
I mean, if the user has a definition like:

```yaml
Sonata\AdminBundle\Tests\App\Admin\FooAdmin:
        arguments: [~, Sonata\AdminBundle\Tests\App\Model\Foo, Sonata\AdminBundle\Controller\CRUDController]
        tags:
            - {name: sonata.admin, manager_type: test, label: Foo}
        calls:
            - [ setFormTheme, [['foo.html.twig' ]]]
```

I guess this `setFormTheme` call from the compiler pass will override the parameters previously set in this definition.

Reply:
With this config, what is called first ?
`setFormTheme, [['foo.html.twig' ]]` or `$definition->addMethodCall('setFormTheme', [$overwriteAdminConfiguration['templates']['form'] ?? []]);` ?

I would expect the `setFormTheme, [['foo.html.twig' ]]` to be called last and override the global configuration. This would allow to remove a form theme and use another one instead for example. Isn't the case ?

But maybe the best would be to expose an `addFormTheme` method instead.

Reply:
I would say `setFormTheme, [['foo.html.twig' ]]` is called first as it is in the definition of the service, and the compiler pass retrieves this definition, but I don't know.

We merge the arguments in the ORM:

https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/7493cfffcbe9996316e2dd0cdd9ac5092cf62a18/src/DependencyInjection/Compiler/AddTemplatesCompilerPass.php#L44

Maybe we can do that.

Reply:
I can have a look today after work. However this would then currently also not work like this because of https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/7493cfffcbe9996316e2dd0cdd9ac5092cf62a18/src/DependencyInjection/Compiler/AddTemplatesCompilerPass.php#L40

So we are not changing this behavior, or?

Reply:
The behavior in doctrineORM is wrong since it's using the sonata-admin configuration.
It will be fixed with https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1427 instead.

Reply:
So I had a look and indeed the call added via `- calls: ...` on the DI service config comes first. So we would overwrite that call.

Shall we support this? Then we have two options:

- merge the method calls
- add a new `addFormTheme(s)` method

Reply:
or we conditionally do this

```php
if (!$definition->hasMethodCall('setFormTheme')) {

Reply:
I guess from a user perspective the most intuitive seems the conditional call, or? Whenever I as a user call `setFormTheme` I would not expect this to be merged with some defaults...

Reply:
Then if the `calls` on the DI service comes first, the issue exist for every method calls.

https://github.com/sonata-project/SonataAdminBundle/blob/d471f78383af57902d54547223291cba8169ac7d/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L351-L356

If we provide an option to override the value, we override the method call without any check.

https://github.com/sonata-project/SonataAdminBundle/blob/d471f78383af57902d54547223291cba8169ac7d/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L341-L348
For all the services, as soon as we have an override, we use it.
For `setFormTheme`, I would say that we're on this case.

But maybe we should modify the code to 
```
if (isset($overwriteAdminConfiguration[$attr]) && !$definition->hasMethodCall($method)) { 
```
instead. This could be considered as a bug.

And then, to be consistent we could add a check
```
if (!$definition->hasMethodCall('setFormTheme')) {
```
in our case. WDYT ?

Reply:
yeah probably this makes the most sense. let me try it

Reply:
Hm if I change it to 

```php
if (isset($overwriteAdminConfiguration[$attr]) && !$definition->hasMethodCall($method)) { 
``` 

then there is a test fail though... not sure we should consider this a bug-fix? it might potentially change things for users out there. Or not?

Reply:
Which one ?

Reply:
```
There was 1 failure:

1) Sonata\AdminBundle\Tests\DependencyInjection\Compiler\AddDependencyCallsCompilerPassTest::testProcessResultingConfig
None of the method calls matched the expected method "setRouteBuilder" with arguments Array &0 (
    0 => 'sonata.admin.route.path_info'
) with any invocation order index
Failed asserting that Symfony\Component\DependencyInjection\Definition Object &000000004bf94b930000000046f6a723 (
    'class' => 'Sonata\AdminBundle\Tests\DependencyInjection\Compiler\MockAdmin'
    'file' => null
    'factory' => null
    'shared' => false
    'deprecation' => Array &0 ()
    'properties' => Array &1 ()
    'calls' => Array &2 (
        0 => Array &3 (
            0 => 'setManagerType'
            1 => Array &4 (
                0 => 'orm'
            )
        )
        1 => Array &5 (
            0 => 'setPagerType'
            1 => Array &6 (
                0 => 'simple'
            )
        )
        2 => Array &7 (
            0 => 'setLabel'
            1 => Array &8 (
                0 => 'Foo'
            )
        )
        3 => Array &9 (
            0 => 'setListModes'
            1 => Array &10 (
                0 => Array &11 (
                    'list' => Array &12 (
                        'class' => 'fa fa-list fa-fw'
                    )
                    'mosaic' => Array &13 (
                        'class' => 'fa fa-th-large fa-fw'
                    )
                )
            )
        )
        4 => Array &14 (
            0 => 'setTemplateRegistry'
            1 => Array &15 (
                0 => Symfony\Component\DependencyInjection\Reference Object &000000004bf94ffb0000000046f6a723 (
                    'id' => 'sonata_news_admin.template_registry'
                    'invalidBehavior' => 1
                )
            )
        )
        5 => Array &16 (
            0 => 'setSecurityInformation'
            1 => Array &17 (
                0 => '%sonata.admin.configuration.security.information%'
            )
        )
        6 => Array &18 (
            0 => 'setFormTheme'
            1 => Array &19 (
                0 => Array &20 ()
            )
        )
        7 => Array &21 (
            0 => 'setFilterTheme'
            1 => Array &22 (
                0 => Array &23 ()
            )
        )
    )
    'instanceof' => Array &24 ()
    'autoconfigured' => false
    'configurator' => null
    'tags' => Array &25 (
        'sonata.admin' => Array &26 (
            0 => Array &27 (
                'group' => 'sonata_group_two'
                'label' => '5 Entry'
                'manager_type' => 'orm'
            )
        )
    )
    'public' => true
    'synthetic' => false
    'abstract' => false
    'lazy' => false
    'decoratedService' => null
    'autowired' => false
    'changes' => Array &28 (
        'public' => true
        'class' => true
        'shared' => true
    )
    'bindings' => Array &29 ()
    'errors' => Array &30 ()
    'arguments' => Array &31 (
        0 => 'sonata_news_admin'
        1 => 'Sonata\AdminBundle\Tests\DependencyInjection\Compiler\News'
        2 => 'Sonata\AdminBundle\Controller\CRUDController'
    )
    'innerServiceId' => null
    'decorationOnInvalid' => null
) has a method call to "setRouteBuilder" with the given arguments..

/var/www/SonataAdminBundle/vendor/matthiasnoback/symfony-dependency-injection-test/PhpUnit/DefinitionHasMethodCallConstraint.php:59
/var/www/SonataAdminBundle/vendor/matthiasnoback/symfony-dependency-injection-test/PhpUnit/AbstractContainerBuilderTestCase.php:188
/var/www/SonataAdminBundle/tests/DependencyInjection/Compiler/AddDependencyCallsCompilerPassTest.php:195
```

Reply:
So we are not setting default service calls then anymore, right? If there is no overwrite and no call yet.

Reply:
maybe like this then? tests pass

```diff
diff --git a/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php b/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php
index b7af06402..c4b4508e9 100644
--- a/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php
+++ b/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php
@@ -338,7 +338,7 @@ class AddDependencyCallsCompilerPass implements CompilerPassInterface
 
             $method = $this->generateSetterMethodName($attr);
 
-            if (isset($overwriteAdminConfiguration[$attr]) || !$definition->hasMethodCall($method)) {
+            if (!$definition->hasMethodCall($method)) {
                 $args = [new Reference($overwriteAdminConfiguration[$attr] ?? $addServiceId)];
                 if ('translator' === $attr) {
                     $args[] = false;
``` 

so this would never overwrite any existing call